README.md
# stacktrace-metadata [![NPM version](https://img.shields.io/npm/v/stacktrace-metadata.svg?style=flat)](https://www.npmjs.com/package/stacktrace-metadata) [![mit license][license-img]][license-url] [![NPM monthly downloads](https://img.shields.io/npm/dm/stacktrace-metadata.svg?style=flat)](https://npmjs.org/package/stacktrace-metadata) [![npm total downloads][downloads-img]][downloads-url]

> Modify given `err` object to be more useful - adds `at`, `line`, `column`, `place` and `filename` properties and also cleans stack traces.

[![code climate][codeclimate-img]][codeclimate-url] 
[![code style][standard-img]][standard-url] 
[![linux build][travis-img]][travis-url] 
[![windows build][appveyor-img]][appveyor-url] 
[![code coverage][coverage-img]][coverage-url] 
[![dependency status][david-img]][david-url]
[![paypal donate][paypalme-img]][paypalme-url] 

You might also be interested in [clean-stacktrace](https://github.com/tunnckocore/clean-stacktrace#readme).

## Highlights

- Works on Windows and Node.js >= 0.10
- Always normalizes Windows paths to be Unix-style
- Customizable through options
- Cleans stack trace using [clean-stacktrace][], `opts.cleanStack: false` to disable
- Passing different current working directory to use, `opts.cwd`, using [clean-stacktrace-relative-paths][]
- Makes paths relative against `opts.cwd`, `opts.relativePaths: false` to disable
- Makes stack trace short - top 4, `opts.shortStack: false` to disable
- Allow hiding the stack trace (making it empty string), `opts.showStack: false` to enable
- Adds useful properties like `line`, `column` and `filename`, using [clean-stacktrace-metadata][]

## Table of Contents
- [Install](#install)
- [Usage](#usage)
- [API](#api)
  * [stacktraceMetadata](#stacktracemetadata)
- [Related](#related)
- [Contributing](#contributing)
- [Building docs](#building-docs)
- [Running tests](#running-tests)
- [Author](#author)
- [License](#license)

_(TOC generated by [verb](https://github.com/verbose/verb) using [markdown-toc](https://github.com/jonschlinkert/markdown-toc))_

## Install
Install with [npm](https://www.npmjs.com/)

```
$ npm install stacktrace-metadata --save
```

or install using [yarn](https://yarnpkg.com)

```
$ yarn add stacktrace-metadata
```

## Usage
> For more use-cases see the [tests](test.js)

```js
const stacktraceMetadata = require('stacktrace-metadata')
```

## API

### [stacktraceMetadata](index.js#L69)
> Cleans stack trace and attaches few more metadata properties, such as `at`, `line`, `column`, `filename` and `place`. By default it cleans stack, makes is short (4 length) and makes paths relative. But all this is controllable through `options` object. Throws `TypeError` if `error` is not an instance of Error.

**Params**

* `error` **{Error}**: real error object, checked against `instanceof Error`    
* `options` **{Object}**: optional options object for more control    
* `options.cleanStack` **{Boolean}**: if `false` won't clean stack trace from node internals    
* `options.shortStack` **{Boolean}**: if `false` full stack traces, otherwise they are just four    
* `options.showStack` **{Boolean}**: if `false` the error.stack will be empty string    
* `options.relativePaths` **{Boolean}**: if `false` paths in stack traces will be absolute    
* `options.mapper` **{Function}**: called on each line of the stack with `(line, index)` signature    
* `options.cwd` **{String}**: current working directory, default `process.cwd()`    
* `returns` **{Error}**: same error object, but modified  

**Example**

```js
const metadata = require('stacktrace-metadata')

const error = new Error('foo quxie')
error.stack = `Error: foo quxie
    at zazz (/home/charlike/apps/alwa.js:8:10)
    at module.exports (/home/charlike/apps/foo.js:6:3)
    at Object.<anonymous> (/home/charlike/apps/dush.js:45:3)
    at Module._compile (module.js:409:26)
    at Object.Module._extensions..js (module.js:416:10)
    at Module.load (module.js:343:32)
    at Function.Module._load (module.js:300:12)
    at Function.Module.runMain (module.js:441:10)
    at startup (node.js:139:18)
`
const err = metadata(error)

console.log(err.line) // => 8
console.log(err.column) // => 10
console.log(err.filename) // => 'alwa.js'
console.log(err.place) // => 'zazz'
console.log(err.at) // => 'zazz (alwa.js:8:10)'
console.log(err.stack)
// =>
// Error: foo quxie
//     at zazz (alwa.js:8:10)
//     at module.exports (foo.js:6:3)
//     at Object.<anonymous> (dush.js:45:3)
```

## Related
- [abbrev-kindof](https://www.npmjs.com/package/abbrev-kindof): Use abbreviations for checking type of given value. Like `kindof(val, 'soa')` to check that value is string, object or array. | [homepage](https://github.com/tunnckocore/abbrev-kindof#readme "Use abbreviations for checking type of given value. Like `kindof(val, 'soa')` to check that value is string, object or array.")
- [assert-kindof](https://www.npmjs.com/package/assert-kindof): Check native type of value and throw AssertionError if not okey. Clean stack traces. Simplicity. Built on [is-kindof][]. | [homepage](https://github.com/tunnckocore/assert-kindof#readme "Check native type of value and throw AssertionError if not okey. Clean stack traces. Simplicity. Built on [is-kindof][].")
- [clean-stack](https://www.npmjs.com/package/clean-stack): Clean up error stack traces | [homepage](https://github.com/sindresorhus/clean-stack#readme "Clean up error stack traces")
- [clean-stacktrace-metadata](https://www.npmjs.com/package/clean-stacktrace-metadata): Plugin for `clean-stacktrace` lib. Parse each line to get additional info like `filename`, `column` and `line` of the error. | [homepage](https://github.com/tunnckocore/clean-stacktrace-metadata#readme "Plugin for `clean-stacktrace` lib. Parse each line to get additional info like `filename`, `column` and `line` of the error.")
- [clean-stacktrace-relative-paths](https://www.npmjs.com/package/clean-stacktrace-relative-paths): Meant to be used with [clean-stacktrace][] as mapper function. Makes absolute paths inside stack traces to relative paths. | [homepage](https://github.com/tunnckocore/clean-stacktrace-relative-paths#readme "Meant to be used with [clean-stacktrace][] as mapper function. Makes absolute paths inside stack traces to relative paths.")
- [clean-stacktrace](https://www.npmjs.com/package/clean-stacktrace): Clean up error stack traces from node internals | [homepage](https://github.com/tunnckocore/clean-stacktrace#readme "Clean up error stack traces from node internals")
- [dush](https://www.npmjs.com/package/dush): Microscopic & functional event emitter in ~260 bytes, extensible through plugins. | [homepage](https://github.com/tunnckocore/dush#readme "Microscopic & functional event emitter in ~260 bytes, extensible through plugins.")
- [is-kindof](https://www.npmjs.com/package/is-kindof): Check type of given javascript value. Support promises, generators, streams, and native types. Built on [kind-of][] lib. | [homepage](https://github.com/tunnckocore/is-kindof#readme "Check type of given javascript value. Support promises, generators, streams, and native types. Built on [kind-of][] lib.")

## Contributing
Pull requests and stars are always welcome. For bugs and feature requests, [please create an issue](https://github.com/tunnckoCore/stacktrace-metadata/issues/new).  
Please read the [contributing guidelines](CONTRIBUTING.md) for advice on opening issues, pull requests, and coding standards.  
If you need some help and can spent some cash, feel free to [contact me at CodeMentor.io](https://www.codementor.io/tunnckocore?utm_source=github&utm_medium=button&utm_term=tunnckocore&utm_campaign=github) too.

**In short:** If you want to contribute to that project, please follow these things

1. Please DO NOT edit [README.md](README.md), [CHANGELOG.md](CHANGELOG.md) and [.verb.md](.verb.md) files. See ["Building docs"](#building-docs) section.
2. Ensure anything is okey by installing the dependencies and run the tests. See ["Running tests"](#running-tests) section.
3. Always use `npm run commit` to commit changes instead of `git commit`, because it is interactive and user-friendly. It uses [commitizen][] behind the scenes, which follows Conventional Changelog idealogy.
4. Do NOT bump the version in package.json. For that we use `npm run release`, which is [standard-version][] and follows Conventional Changelog idealogy.

Thanks a lot! :)

## Building docs
Documentation and that readme is generated using [verb-generate-readme][], which is a [verb][] generator, so you need to install both of them and then run `verb` command like that

```
$ npm install verbose/verb#dev verb-generate-readme --global && verb
```

_Please don't edit the README directly. Any changes to the readme must be made in [.verb.md](.verb.md)._

## Running tests
Clone repository and run the following in that cloned directory

```
$ npm install && npm test
```

## Author
**Charlike Mike Reagent**

+ [github/tunnckoCore](https://github.com/tunnckoCore)
+ [twitter/tunnckoCore](https://twitter.com/tunnckoCore)
+ [codementor/tunnckoCore](https://codementor.io/tunnckoCore)

## License
Copyright © 2015, 2017, [Charlike Mike Reagent](https://i.am.charlike.online). Released under the [MIT License](LICENSE).

***

_This file was generated by [verb-generate-readme](https://github.com/verbose/verb-generate-readme), v0.4.3, on March 15, 2017._  
_Project scaffolded using [charlike][] cli._

[charlike]: https://github.com/tunnckocore/charlike
[clean-stacktrace-metadata]: https://github.com/tunnckocore/clean-stacktrace-metadata
[clean-stacktrace-relative-paths]: https://github.com/tunnckocore/clean-stacktrace-relative-paths
[clean-stacktrace]: https://github.com/tunnckocore/clean-stacktrace
[commitizen]: https://github.com/commitizen/cz-cli
[is-kindof]: https://github.com/tunnckocore/is-kindof
[kind-of]: https://github.com/jonschlinkert/kind-of
[standard-version]: https://github.com/conventional-changelog/standard-version
[verb-generate-readme]: https://github.com/verbose/verb-generate-readme
[verb]: https://github.com/verbose/verb

[license-url]: https://www.npmjs.com/package/stacktrace-metadata
[license-img]: https://img.shields.io/npm/l/stacktrace-metadata.svg

[downloads-url]: https://www.npmjs.com/package/stacktrace-metadata
[downloads-img]: https://img.shields.io/npm/dt/stacktrace-metadata.svg

[codeclimate-url]: https://codeclimate.com/github/tunnckoCore/stacktrace-metadata
[codeclimate-img]: https://img.shields.io/codeclimate/github/tunnckoCore/stacktrace-metadata.svg

[travis-url]: https://travis-ci.org/tunnckoCore/stacktrace-metadata
[travis-img]: https://img.shields.io/travis/tunnckoCore/stacktrace-metadata/master.svg?label=linux

[appveyor-url]: https://ci.appveyor.com/project/tunnckoCore/stacktrace-metadata
[appveyor-img]: https://img.shields.io/appveyor/ci/tunnckoCore/stacktrace-metadata/master.svg?label=windows

[coverage-url]: https://codecov.io/gh/tunnckoCore/stacktrace-metadata
[coverage-img]: https://img.shields.io/codecov/c/github/tunnckoCore/stacktrace-metadata/master.svg

[david-url]: https://david-dm.org/tunnckoCore/stacktrace-metadata
[david-img]: https://img.shields.io/david/tunnckoCore/stacktrace-metadata.svg

[standard-url]: https://github.com/feross/standard
[standard-img]: https://img.shields.io/badge/code%20style-standard-brightgreen.svg

[paypalme-url]: https://www.paypal.me/tunnckoCore
[paypalme-img]: https://img.shields.io/badge/paypal-donate-brightgreen.svg



CHANGELOG.md
# Change Log

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

<a name="2.0.4"></a>
## [2.0.4](https://github.com/tunnckoCore/stacktrace-metadata/compare/v2.0.3...v2.0.4) (2017-03-19)


### Bug Fixes

* **deps:** update deps, again! ([558d43d](https://github.com/tunnckoCore/stacktrace-metadata/commit/558d43d))



<a name="2.0.3"></a>
## [2.0.3](https://github.com/tunnckoCore/stacktrace-metadata/compare/v2.0.2...v2.0.3) (2017-03-19)


### Bug Fixes

* **package:** force update deps ([8ec3b33](https://github.com/tunnckoCore/stacktrace-metadata/commit/8ec3b33))



<a name="2.0.2"></a>
## [2.0.2](https://github.com/tunnckoCore/stacktrace-metadata/compare/v2.0.1...v2.0.2) (2017-03-18)


### Bug Fixes

* **deps:** update deps ([b71f81e](https://github.com/tunnckoCore/stacktrace-metadata/commit/b71f81e))
* **package:** update deps and better metadata mapper ([00502c5](https://github.com/tunnckoCore/stacktrace-metadata/commit/00502c5))



<a name="2.0.1"></a>
## [2.0.1](https://github.com/tunnckocore/stacktrace-metadata/compare/v2.0.0...v2.0.1) (2017-03-15)


### Bug Fixes

* **ci:** fix tests for CI ([ad7cd78](https://github.com/tunnckocore/stacktrace-metadata/commit/ad7cd78))
* **detection:** fix bug in detecting the startStackFunction correctly ([e5e3b34](https://github.com/tunnckocore/stacktrace-metadata/commit/e5e3b34)), closes [#5](https://github.com/tunnckocore/stacktrace-metadata/issues/5)
* **options:** introduce custom opts.mapper function ([66bb5f5](https://github.com/tunnckocore/stacktrace-metadata/commit/66bb5f5)), closes [#3](https://github.com/tunnckocore/stacktrace-metadata/issues/3)
* **package:** update deps ([701e786](https://github.com/tunnckocore/stacktrace-metadata/commit/701e786))
* **tests:** improve test ([a03c703](https://github.com/tunnckocore/stacktrace-metadata/commit/a03c703)), closes [#4](https://github.com/tunnckocore/stacktrace-metadata/issues/4)
* **tests:** update tests to not fail on CI cov ([8413c8c](https://github.com/tunnckocore/stacktrace-metadata/commit/8413c8c))



<a name="2.0.0"></a>
# [2.0.0](https://github.com/tunnckocore/stacktrace-metadata/compare/v1.0.1...v2.0.0) (2017-03-13)


### Bug Fixes

* **index:** micro optimizations ([ad1c032](https://github.com/tunnckocore/stacktrace-metadata/commit/ad1c032))
* **package:** add keywords ([ff84b05](https://github.com/tunnckocore/stacktrace-metadata/commit/ff84b05))
* **package:** bump deps ([e8021dd](https://github.com/tunnckocore/stacktrace-metadata/commit/e8021dd))
* **revert:** micro optimizations are not good thing some times ([66fd63a](https://github.com/tunnckocore/stacktrace-metadata/commit/66fd63a))


### Code Refactoring

* **rewrite:** from scratch, update boilerplate ([49e95e7](https://github.com/tunnckocore/stacktrace-metadata/commit/49e95e7))


### BREAKING CHANGES

* **rewrite:** Now uses more low-level libs and has options object





## 1.0.1 - 2016-02-17
- Release v1.0.1 / npm@v1.0.1
- update deps

## 1.0.0 - 2016-02-13
- Release v1.0.0 / npm@v1.0.0
- official v1 release, stable forever
- update boilerplate
- add related libs
- add example
- add tests

## 0.0.0 - 2015-09-11
- Initial commit

test.js
/*!
 * stacktrace-metadata <https://github.com/tunnckoCore/stacktrace-metadata>
 *
 * Copyright (c) Charlike Mike Reagent <@tunnckoCore> (https://i.am.charlike.online)
 * Released under the MIT license.
 */

/* jshint asi:true */

'use strict'

var test = require('mukla')
var stacktraceMetadata = require('./index')

test('should throw TypeError if not passed with real error', function (done) {
  function fixture () {
    stacktraceMetadata(123)
  }

  test.throws(fixture, TypeError)
  test.throws(fixture, /expect `error` to be real error/)
  done()
})

test('should return the same error', function firstTest (done) {
  var err = new Error('foo bar')
  var e = stacktraceMetadata(err)

  test.strictEqual(err === e, true)
  test.strictEqual(e instanceof Error, true)
  test.strictEqual(err.name, e.name)
  test.strictEqual(err.stack, e.stack)
  test.strictEqual(err.message, e.message)
  test.strictEqual(err.message, e.message)
  test.strictEqual(e.line, 26)
  test.ok(e.column)
  test.strictEqual(e.filename, 'test.js')
  test.strictEqual(e.place, 'Object.firstTest')
  done()
})

test('should have relative paths by default', function (done) {
  var err = new Error('opts.relativePaths: true')
  var e = stacktraceMetadata(err)

  test.strictEqual(e.message, 'opts.relativePaths: true')
  test.strictEqual(e.stack.indexOf('(test.js:43:13') > 0, true)
  test.strictEqual(e.line, 43)
  test.strictEqual(e.filename, 'test.js')
  test.ok(e.at)
  test.ok(e.column)
  done()
})

test('should have absolute paths if opts.relativePaths: false', function (done) {
  var error = new Error('foo qux bar')
  var e = stacktraceMetadata(error, {
    relativePaths: false
  })

  test.strictEqual(e.message, 'foo qux bar')
  test.strictEqual(e.stack.indexOf('(test.js:') === -1, true)
  test.strictEqual(e.line, 56)
  test.ok(e.at)
  test.ok(e.place)
  test.ok(e.column)
  test.ok(e.filename)
  done()
})

test('should clean stack by default', function (done) {
  var error = new TypeError('woohooo')
  var stack = [
    'Error: woohooo',
    '    at Object.xyz (/home/charlike/apps/stacktrace-metadata/test.js:111:33)',
    '    at Object.tryCatch (/home/charlike/apps/node_modules/try-catch-callback/index.js:75:14)',
    '    at Object.tryCatchCallback (/home/charlike/apps/node_modules/try-catch-callback/index.js:58:21)',
    '    at Object.tryCatch (/home/charlike/apps/node_modules/always-done/node_modules/try-catch-core/index.js:80:26)',
    '    at Object.tryCatchCore (/home/charlike/apps/node_modules/always-done/node_modules/try-catch-core/index.js:64:12)',
    '    at Object.alwaysDone (/home/charlike/apps/node_modules/always-done/index.js:61:24)',
    '    at mukla (/home/charlike/apps/node_modules/mukla/index.js:55:9)',
    '    at Object.<anonymous> (/home/charlike/apps/stacktrace-metadata/zazzy.js:15:1)',
    '    at Module._compile (module.js:571:32)',
    '    at Object.Module._extensions..js (module.js:580:10)'
  ]
  error.stack = stack.join('\n')

  var e = stacktraceMetadata(error, {
    shortStack: false
  })

  // should not exist, because stack is cleaned
  var notExists = e.stack.indexOf('at Module._compile (module.js:571:32)') === -1
  test.strictEqual(notExists, true)

  // should new stack be shorter than the old one
  test.strictEqual(e.stack.split('\n').length < stack.length, true)
  test.ok(e.line)
  test.strictEqual(e.place, 'Object.xyz')
  test.strictEqual(/Object\.xyz/.test(e.at), true)
  test.strictEqual(/test\.js:111:33/.test(e.at), true)
  test.ok(e.filename)
  test.ok(e.column)
  done()
})

test('should not clean stack if opts.cleanStack: false', function bazzyTest (done) {
  var err = new Error('quxie bazzy')
  var opts = {
    cleanStack: false,
    shortStack: false
  }
  var e = stacktraceMetadata(err, opts)

  var internals = e.stack.indexOf('at Module._compile') > 0
  test.strictEqual(internals, true)
  test.strictEqual(e.line, 108)
  test.ok(e.filename)
  test.strictEqual(e.place, 'Object.bazzyTest')
  test.strictEqual(/Object\.bazzyTest/.test(e.at), true)
  test.strictEqual(/test\.js:108:13/.test(e.at), true)
  test.ok(e.column)
  done()
})

test('should have empty string err.stack property if opts.showStack: false', function emptyStack (done) {
  var err = stacktraceMetadata(new Error('abc'), {
    showStack: false
  })

  test.strictEqual(err.stack, '')
  test.strictEqual(err.line, 127)
  test.strictEqual(err.place, 'Object.emptyStack')
  test.strictEqual(/test\.js:127:32/.test(err.at), true)
  test.ok(err.filename)
  test.ok(err.column)
  done()
})

test('should have props like `err.line`, `err.filename` and `err.column`', function myQuxTest (done) {
  var err = new Error('my special error')
  var e = stacktraceMetadata(err)

  test.strictEqual(e.name, 'Error')
  test.strictEqual(e.message, 'my special error')
  test.strictEqual(e.line, 141)
  test.strictEqual(e.place, 'Object.myQuxTest')
  test.ok(e.filename)
  test.strictEqual(/Object\.myQuxTest/.test(e.at), true)
  test.strictEqual(/test\.js:141:13/.test(e.at), true)
  test.ok(err.column)
  done()
})

test('should not have special props if passed error has empty stack', function (done) {
  var err = new Error('foo')
  err.stack = ''

  var e = stacktraceMetadata(err)

  test.strictEqual(e.at, undefined)
  test.strictEqual(e.line, undefined)
  test.strictEqual(e.place, undefined)
  test.strictEqual(e.column, undefined)
  test.strictEqual(e.filename, undefined)
  test.strictEqual(e === err, true)
  done()
})

test('should work for errors thrown like what rimraf.sync throws', function quxieTest (done) {
  var rimraf = require('rimraf')

  try {
    rimraf.sync(12345)
  } catch (err) {
    var e = stacktraceMetadata(err, {
      mapper: function () {}
    })

    test.strictEqual(e.line, 174)
    test.strictEqual(e.place, 'Object.quxieTest')
    test.ok(e.filename)
    test.strictEqual(/Object\.quxieTest/.test(e.at), true)
    test.strictEqual(/test\.js:174:12/.test(e.at), true)
    test.ok(err.column)
  }
  done()
})

test('should be able to pass custom opts.mapper function', function (done) {
  var err = new Error('custom mapper')
  var e = stacktraceMetadata(err, {
    mapper: function (line) {
      test.strictEqual(typeof line, 'string')
      return line
    }
  })

  test.strictEqual(e.stack.trim().length > 0, true)
  test.ok(e.at)
  test.ok(e.line)
  test.ok(e.place)
  test.ok(e.column)
  test.ok(e.filename)
  done()
})

test('should work for stack that dont have "place" at some lines', function (done) {
  var msg = 'no place just filename, line and column'
  var err = new Error(msg)
  err.stack = [
    'Error: no place just filename, line and column',
    '    at /home/foo/bar.js:331:7'
  ].join('\n')

  var error = stacktraceMetadata(err)
  test.strictEqual(error.place, '')
  done()
})


index.js
/*!
 * stacktrace-metadata <https://github.com/tunnckoCore/stacktrace-metadata>
 *
 * Copyright (c) Charlike Mike Reagent <@tunnckoCore> (https://i.am.charlike.online)
 * Released under the MIT license.
 */

'use strict'

var extend = require('extend-shallow')
var findCallsite = require('find-callsite')
var clean = require('clean-stacktrace')
var metadata = require('clean-stacktrace-metadata')
var relativePaths = require('clean-stacktrace-relative-paths')

/**
 * > Cleans stack trace and attaches few more metadata properties,
 * such as `at`, `line`, `column`, `filename` and `place`. By default
 * it cleans stack, makes is short (4 length) and makes paths relative.
 * But all this is controllable through `options` object.
 * Throws `TypeError` if `error` is not an instance of Error.
 *
 * **Example**
 *
 * ```js
 * const metadata = require('stacktrace-metadata')
 *
 * const error = new Error('foo quxie')
 * error.stack = `Error: foo quxie
 *     at zazz (/home/charlike/apps/alwa.js:8:10)
 *     at module.exports (/home/charlike/apps/foo.js:6:3)
 *     at Object.<anonymous> (/home/charlike/apps/dush.js:45:3)
 *     at Module._compile (module.js:409:26)
 *     at Object.Module._extensions..js (module.js:416:10)
 *     at Module.load (module.js:343:32)
 *     at Function.Module._load (module.js:300:12)
 *     at Function.Module.runMain (module.js:441:10)
 *     at startup (node.js:139:18)
 * `
 * const err = metadata(error)
 *
 * console.log(err.line) // => 8
 * console.log(err.column) // => 10
 * console.log(err.filename) // => 'alwa.js'
 * console.log(err.place) // => 'zazz'
 * console.log(err.at) // => 'zazz (alwa.js:8:10)'
 * console.log(err.stack)
 * // =>
 * // Error: foo quxie
 * //     at zazz (alwa.js:8:10)
 * //     at module.exports (foo.js:6:3)
 * //     at Object.<anonymous> (dush.js:45:3)
 * ```
 *
 * @name   stacktraceMetadata
 * @param  {Error} `error` real error object, checked against `instanceof Error`
 * @param  {Object} `options` optional options object for more control
 * @param  {Boolean} `options.cleanStack` if `false` won't clean stack trace from node internals
 * @param  {Boolean} `options.shortStack` if `false` full stack traces, otherwise they are just four
 * @param  {Boolean} `options.showStack` if `false` the error.stack will be empty string
 * @param  {Boolean} `options.relativePaths` if `false` paths in stack traces will be absolute
 * @param  {Function} `options.mapper` called on each line of the stack with `(line, index)` signature
 * @param  {String} `options.cwd` current working directory, default `process.cwd()`
 * @return {Error} same error object, but modified
 * @throws {TypeError} If `error` not instance of Error
 * @api public
 */

module.exports = function stacktraceMetadata (error, options) {
  if (!(error instanceof Error)) {
    throw new TypeError('stacktrace-metadata: expect `error` to be real error')
  }

  if (typeof error.stack === 'string' && error.stack.length) {
    var opts = extend({
      showStack: true,
      shortStack: true,
      cleanStack: true,
      relativePaths: true
    }, options)

    var relative = function relative (val) {
      return opts.relativePaths ? relativePaths(opts.cwd)(val) : val
    }

    var stack = clean(error.stack, function mapper (line, index) {
      line = typeof opts.mapper === 'function'
        ? (opts.mapper(line, index) || line)
        : line
      line = relative(line)
      return line
    })

    if (!opts.cleanStack) {
      stack = error.stack
    }

    var at = findCallsite(stack, opts)
    metadata(mapdata(error))(at)

    stack = opts.shortStack
      ? stack.split('\n').splice(0, 4).join('\n')
      : stack

    error.stack = opts.showStack ? stack : ''
    return error
  }

  return error
}

function mapdata (error) {
  return function mapper (_, info) {
    var atLine = info.place.length ? [
      info.place,
      ' ('
    ] : []

    atLine = atLine.concat([
      info.filename,
      ':',
      info.line,
      ':',
      info.column
    ])

    atLine = info.place.length ? atLine.concat(')') : atLine
    error.at = atLine.join('')

    error.line = info.line
    error.place = info.place
    error.column = info.column
    error.filename = info.filename
  }
}
